presto-main/src/main/java/com/facebook/presto/execution/scheduler/NodeScheduler.java

```diff
@@ -306,40 +309,82 @@ public static int calculateLowWatermark(int maxPendingSplitsPerTask)
+
+        boolean blocked = true;
+        ImmutableList.Builder<FutureStateChange<?>.Listener> listeners = ImmutableList.builder();
+        for (Node blockedNode : blockedNodes) {
+            RemoteTask remoteTask = nodeToTaskMap.get(blockedNode.getNodeIdentifier());
+            if (remoteTask == null) {
+                continue;
+            }
+            Optional<FutureStateChange<?>.Listener> listener = remoteTask.whenSplitQueueHasSpace(spaceThreshold);
+            if (!listener.isPresent()) {
+                blocked = false;
+                break;
+            }
+            listeners.add(listener.get());
+        }
+        if (!blocked) {
             return immediateFuture(null);
         }
-        return getFirstCompleteAndCancelOthers(blockedFutures);
+        return getFirstCompleteAndCancelOthers(listeners.build());
     }
 
     public static ListenableFuture<?> toWhenHasSplitQueueSpaceFuture(List<RemoteTask> existingTasks, int spaceThreshold)
     {
         if (existingTasks.isEmpty()) {
             return immediateFuture(null);
         }
-        List<ListenableFuture<?>> stateChangeFutures = existingTasks.stream()
-                .map(remoteTask -> remoteTask.whenSplitQueueHasSpace(spaceThreshold))
-                .collect(toImmutableList());
-        return getFirstCompleteAndCancelOthers(stateChangeFutures);
+
+        boolean blocked = true;
+        ImmutableList.Builder<FutureStateChange<?>.Listener> listeners = ImmutableList.builder();
+        for (RemoteTask remoteTask : existingTasks) {
+            Optional<FutureStateChange<?>.Listener> listener = remoteTask.whenSplitQueueHasSpace(spaceThreshold);
+            if (!listener.isPresent()) {
+                blocked = false;
+                break;
+            }
+            listeners.add(listener.get());
+        }
+        if (!blocked) {
+            return immediateFuture(null);
+        }
+        return getFirstCompleteAndCancelOthers(listeners.build());
     }
 
-    private static ListenableFuture<?> getFirstCompleteAndCancelOthers(List<ListenableFuture<?>> blockedFutures)
+    /**
+     * Wait for any in the listeners to fire, and cancel all others to free up resources
+     */
+    private static ListenableFuture<?> getFirstCompleteAndCancelOthers(List<FutureStateChange<?>.Listener> listeners)
     {
-        // wait for the first task to unblock and then cancel all futures to free up resources
-        ListenableFuture<?> result = whenAnyComplete(blockedFutures);
-        result.addListener(
-                () -> {
-                    for (ListenableFuture<?> blockedFuture : blockedFutures) {
-                        blockedFuture.cancel(true);
-                    }
-                },
-                MoreExecutors.directExecutor());
-        return result;
+        SettableFuture<?> future = SettableFuture.create();
+
+        // make sure consumer callback does not run before the completion of this method
```

Comment:
Can you add to this comment why we need this guarantee (e.g., what happens if the callback runs before the completion of this method?)?